datalad_next/annexremote/uncurl.py

```diff
@@ -0,0 +1,349 @@
+- `git_remotename` - Name of the Git remote for the uncurl special remote
+
+Additional components can be defined by configuring `match` expression
+configurations that are evaluated on each recorded URL for any given git-annex
+key. It is recommended that each `match` expression uses unique match group
+identifiers to avoid value conflicts.
+
+https://docs.python.org/3/library/string.html#format-string-syntax
+"""
+
+from functools import partial
+import json
+from pathlib import Path
+import re
+from urllib.parse import urlparse
+
+# we intentionally limit ourselves to the most basic interface
+# and even that we only need to get a `ConfigManager` instance.
+# If that class would support a plain path argument, we could
+# avoid it entirely
+from datalad.dataset.gitrepo import GitRepo
+
+from datalad_next.exceptions import (
+    CapturedException,
+    DownloadError,
+)
+from datalad_next.url_operations.any import AnyUrlOperations
+from datalad_next.utils import ensure_list
+
+from . import (
+    RemoteError,
+    SpecialRemote,
+    UnsupportedRequest,
+    super_main
+)
+
+
+
+class UncurlRemote(SpecialRemote):
+    """ """
+    def __init__(self, annex):
+        super().__init__(annex)
+        self.configs.update(
+            url='Python format language template composing an access URL',
+            match='(whitespace-separated list of) regular expression(s) to match particular components in supported URL via named groups',
+        )
+        self.repo = None
+        self.url_tmpl = None
+        self.match = None
+        self.url_handler = None
+        # cache of properties that do not vary within a session
+        # or across annex keys
+        self.persistent_tmpl_props = {}
+
+    def initremote(self):
+        # at present there is nothing that needs to be done on init/enable.
+        # the remote is designed to work without any specific setup too
+        pass
+
+    def prepare(self):
+        # we need the git remote name to be able to look up config about
+        # that remote
+        remotename = self.annex.getgitremotename()
+        # get the repo to gain access to its config
+        self.repo = GitRepo(self.annex.getgitdir())
+        # check the config for a URL template setting
+        self.url_tmpl = self.repo.cfg.get(
+            f'remote.{remotename}.uncurl-url', '')
+        # only if we have no local, overriding, configuration ask git-annex
+        # for the committed special remote config on the URL template
+        if not self.url_tmpl:
+            # ask for the commit config, could still be empty
+            self.url_tmpl = self.annex.getconfig('url')
+            # TODO test the case of a fully absent URL template
+            # that would be fine and only verbatim recorded URLs could
+            # be sent to a downloader
+
+        # unconditionally ask git-annex for a match-url setting, any local
+        # config ammends, and does not override
```

Comment:
If understand correctly the rational here is, that while I may have a better idea about the to-be-formed URL locally, the same would not be true for "known" URL patterns - I may know additional ones, but that doesn't invalidate already known ones. Hence, no override but only amendment. 

However, this needs clear documentation since the difference in behavior WRT local configs can be quite surprising. (All the more, if my understanding of the rationale was not correct)

Reply:
I have added a section on configuration to the docs.